virtual_rainforest/core/config.py

```diff
@@ -315,11 +386,42 @@ def construct_combined_schema(modules: list[str]) -> dict:
     return comb_schema
 
 
+def validate_with_defaults(config_dict: dict, comb_schema: dict) -> dict:
+    """Validate the configuration settings against the combined schema.
+
+    This function also adds default values into the configuration dictionary where it is
+    appropriate.
+
+     Args:
+        config_dict: The complete configuration settings for the particular model
+            instance
+        comb_schema: Combined schema for all modules that are being configured
+
+    Raises:
+        RuntimeError: If the configuration files fail to validate against the JSON
+            schema
+    """
+
+    # Make a new validator that allows the addition of defaults
+    ValidatorWithDefaults = extend_with_default(Draft202012Validator)
+
+    # Validate the input configuration settings against the combined schema
+    # This step also adds in all default module configuration details
+    try:
+        ValidatorWithDefaults(comb_schema).validate(config_dict)
+    except exceptions.ValidationError as err:
+        log_and_raise(
+            f"Validation of configuration files failed: {err.message}", RuntimeError
+        )
+
+    return config_dict
+
+
 def validate_config(
     cfg_paths: Union[str, list[str]],
```

Comment:
You're using Python 3.9 right? If you're using Python 3.10 you can write this as `str | list[str]`

Reply:
Yeah we are, shame as it's a much nicer syntax